idem_azurerm/exec/azurerm/application_insights/component.py

```diff
@@ -0,0 +1,299 @@
+            "applicationinsights", str(exc), **kwargs
+        )
+        result = {"error": str(exc)}
+    except ValidationError as exc:
+        result = {"error": str(exc)}
+
+    return result
+
+
+async def delete(
+    hub, ctx, name, resource_group, **kwargs,
+):
+    """
+    .. versionadded:: 3.0.0
+
+    Deletes an Application Insights component.
+
+    :param name: The name of the component to delete.
+
+    :param resource_group: The name of the resource group.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        azurerm.application_insights.component.delete test_name test_group
+
+    """
+    result = False
+    insightconn = await hub.exec.azurerm.utils.get_client(
+        ctx, "applicationinsights", **kwargs
+    )
+
+    try:
+        component = insightconn.components.delete(
+            resource_name=name, resource_group_name=resource_group,
+        )
+
+        result = True
+    except CloudError as exc:
+        await hub.exec.azurerm.utils.log_cloud_error(
+            "applicationinsights", str(exc), **kwargs
+        )
+        result = {"error": str(exc)}
+
+    return result
+
+
+async def get(hub, ctx, name, resource_group, **kwargs):
+    """
+    .. versionadded:: 3.0.0
+
+    Returns an Application Insights component.
+
+    :param name: The name of the component.
+
+    :param resource_group: The name of the resource group.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        azurerm.application_insights.component.get test_name test_group
+
+    """
+    result = {}
+    insightconn = await hub.exec.azurerm.utils.get_client(
+        ctx, "applicationinsights", **kwargs
+    )
+
+    try:
+        component = insightconn.components.get(
+            resource_name=name, resource_group_name=resource_group
+        )
+
+        result = component.as_dict()
+    except CloudError as exc:
+        await hub.exec.azurerm.utils.log_cloud_error(
+            "applicationinsights", str(exc), **kwargs
+        )
+        result = {"error": str(exc)}
+
+    return result
+
+
+async def list_(hub, ctx, **kwargs):
+    """
+    .. versionadded:: 3.0.0
+
+    Gets a list of all Application Insights components within a subscription.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        azurerm.application_insights.component.list
+
+    """
+    result = {}
+    insightconn = await hub.exec.azurerm.utils.get_client(
+        ctx, "applicationinsights", **kwargs
+    )
+
+    try:
+        components = await hub.exec.azurerm.utils.paged_object_to_list(
+            insightconn.components.list()
+        )
+
+        for component in components:
+            result[component["name"]] = component
+    except CloudError as exc:
+        await hub.exec.azurerm.utils.log_cloud_error(
+            "applicationinsights", str(exc), **kwargs
+        )
+        result = {"error": str(exc)}
+
+    return result
+
+
+async def list_by_resource_group(hub, ctx, resource_group, **kwargs):
```

Comment:
take a look at what I've been doing recently with the container `list*` modules. I have a single `list` function with an optional `resource_group` parameter. if it's provided, I switch to the `list_by_resource_group` SDK function and abstract it from the user.